pkg/addonmanager/controllers/agentdeploy/controller.go

```diff
@@ -150,20 +151,14 @@ func (c *addonDeployController) sync(ctx context.Context, syncCtx factory.SyncCo
 	}
 
 	cluster, err := c.managedClusterLister.Get(clusterName)
-	if errors.IsNotFound(err) {
-		// to delete the addon to make sure the addon is cleaned up when the cluster is not found.
-		// need go to sync to handle cleanup if the addon is deleting.
-		// in the normal case the addon should have been cleaned up after the cluster is deleted.
-		if addon.DeletionTimestamp.IsZero() {
-			return c.cleanupAddon(ctx, clusterName, addonName)
-		}
-	}
 	if err != nil {
-		return err
+		// TODO: should consider to force delete the addon and its manifestWork if cluster is not found.
+		return client.IgnoreNotFound(err)
 	}
 
-	if !cluster.DeletionTimestamp.IsZero() {
-		// TODO: consider to delete addon here
+	// waiting for the addon is deleted after cluster is deleted.
+	// need go to sync to handle cleanup if the addon is deleting.
+	if !cluster.DeletionTimestamp.IsZero() && addon.DeletionTimestamp.IsZero() {
```

Comment:
this is also confusing, why you need to care addon status.

Or you should care about cluster status in each syncer instead?

Reply:
moved the cluster status check to each syncer.

when the cluster is deleting, there are 2 cases:
1. the addon is not deleting, we should not apply/update the deploy manifestWork, and wait until the addon is deleted.
2. the addon is deleting, we begin the cleanup process to remove the deploy manifestWork.